### PR TITLE
VB Remove Unnecessary Cast should not remove necessary casts from Nothing to T?

### LINKED ISSUE #5685: Conversion of Nothing literal to Date? as operand of ternary If expression incorrectly reported as unnecessary
```VB.NET
Module Module1
    Sub Main()
        Dim d As Date? = If(True, DirectCast(Nothing, Date?), CDate(""))
    End Sub
End Module
```

The DirectCast above is faded. This also affects CType. Removing the conversion would cause the Nothing literal to be interpreted as #1/1/0001 00:00:00# rather than as a null nullable Date.

### PR DESCRIPTION
**Pull request by DustinCampbell**
Fixes #5685

There are several scenarios where removing a cast from Nothing to T? can result in the value of Nothing changing. Essentially, the value uses its "nullness" and gets the default value of T. For example:

```VB
Dim x As Integer? = If(True, DirectCast(Nothing, Integer?), 42)
Console.Write(x)
```

In the code above, ```Write``` won't print anything to the console because the argument passed in is null. However, if the cast is removed, the program changes because ```Nothing``` gets the default value of Integer, which is 0.

This change adds logic for the Nothing to T? case. Additionally, there is now code to compute the outer type for checking casts as the "other side" (i.e. the when-true and when-false) branches of a ternary conditional expression.

Tagging @dotnet/mlangide

### CODE REVIEW

File: src/Workspaces/VisualBasic/Portable/Extensions/CastAnalyzer.vb [hunk 1]
```diff
@@ -373,16 +382,35 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
...
+
+                If castType.Equals(outerType) OrElse outerType.SpecialType = SpecialType.System_Object Then
+                    Return False
+                End If
+
+                Return True
+            End If
+
+            Select Case castType.SpecialType
+                Case SpecialType.System_DateTime
```
Comment (Pilchie) [inline src/Workspaces/VisualBasic/Portable/Extensions/CastAnalyzer.vb]: Is it really specifc types like this, or should we be looking at all value types here?

Comment (DustinCampbell) [inline src/Workspaces/VisualBasic/Portable/Extensions/CastAnalyzer.vb]: I'm not sure. I wanted to do the least amount of changes in here. However, I suspect other cases are handled by the case at the top of the method. I'll go ahead and clean up this method to make that clearer.

**Comment by jasonmalinowski** [comment]
:+1:

**Comment by DustinCampbell** [comment]
test linux please

**Comment by DustinCampbell** [comment]
linux test failure isn't my fault ("process hung")

**Pull request merged**